Core/src/Geometry/TrackingVolume.cpp

```diff
@@ -472,164 +468,158 @@ void Acts::TrackingVolume::closeGeometry(
-        // layer on approach intersection
-        auto atIntersection =
-            tLayer->surfaceOnApproach(gctx, position, direction, options);
-        auto path = atIntersection.pathLength();
-        bool withinLimit = std::abs(path) <= std::abs(options.farLimit);
-        // Intersection is ok - take it (move to surface on approach)
-        if (atIntersection && withinLimit) {
-          // create a layer intersection
-          lIntersections.push_back(LayerIntersection(atIntersection, tLayer));
-        }
+  if (m_confinedLayers == nullptr) {
+    return {};
+  }
+
+  // start layer given or not - test layer
+  const Layer* tLayer = options.startObject != nullptr
+                            ? static_cast<const Layer*>(options.startObject)
+                            : associatedLayer(gctx, position);
+  while (tLayer != nullptr) {
+    // check if the layer needs resolving
+    // - resolveSensitive -> always take layer if it has a surface array
+    // - resolveMaterial -> always take layer if it has material
+    // - resolvePassive -> always take, unless it's a navigation layer
+    // skip the start object
+    if (tLayer != options.startObject && tLayer->resolve(options)) {
+      // if it's a resolveable start layer, you are by definition on it
+      // layer on approach intersection
+      auto atIntersection =
+          tLayer->surfaceOnApproach(gctx, position, direction, options);
+      // Intersection is ok - take it (move to surface on approach)
+      if (atIntersection) {
+        // create a layer intersection
+        lIntersections.push_back(LayerIntersection(atIntersection, tLayer));
       }
-      // move to next one or break because you reached the end layer
-      tLayer = (tLayer == options.endObject)
-                   ? nullptr
-                   : tLayer->nextLayer(gctx, position, direction);
     }
-    std::sort(lIntersections.begin(), lIntersections.end(),
-              [](const LayerIntersection& a, const LayerIntersection& b) {
-                return SurfaceIntersection::forwardOrder(a.first, b.first);
-              });
+    // move to next one or break because you reached the end layer
+    tLayer = (tLayer == options.endObject)
+                 ? nullptr
+                 : tLayer->nextLayer(gctx, position, direction);
   }
-  // and return
+
+  // In case of cylindrical layers we might resolve far intersection solutions
+  // which are not valid for navigation. These are discarded here by checking
+  // against the minimum path length.
+  auto min = std::min_element(
+      lIntersections.begin(), lIntersections.end(),
+      [](const LayerIntersection& a, const LayerIntersection& b) {
+        return a.first.pathLength() < b.first.pathLength();
+      });
+  std::rotate(lIntersections.begin(), min, lIntersections.end());
+  lIntersections.resize(std::distance(min, lIntersections.end()),
+                        {SurfaceIntersection::invalid(), nullptr});
```

Comment:
This rotates the min element to the start and then resizes from min element to the end. But is the min iterative even valid after the rotation? 

I guess this works because it's only ever 1 or 2 elements, right?

Reply:
in some rare cases we resolve a layer on the other side of the cylinder which then messes with the navigation. here I check if the order makes sense and remove the ones which are out of order at the beginning of the vector

Reply:
at the moment they have to be thrown out. otherwise we end up stepping backward since I killed to overstepping limit completely.